module/spock/spock_verifier.go

```diff
@@ -0,0 +1,191 @@
+func (v *Verifier) VerifyApproval(approval *flow.ResultApproval) (*flow.ExecutionReceipt, error) {
+
+	// check if verification already done and that a match is there
+	// note: we do not add to cache if the approval was not verified
+	receipt, ok := v.approvals[approval.Body.ExecutionResultID][approval.ID()]
+	if ok {
+		return receipt, nil
+	}
+
+	// find identities
+	approver, err := v.protocolState.AtBlockID(approval.Body.BlockID).Identity(approval.Body.ApproverID)
+	if err != nil {
+		return nil, fmt.Errorf("could not find approver identity")
+	}
+
+	receipts := v.receipts[approval.Body.ExecutionResultID]
+	for _, receipt := range receipts {
+		executor, err := v.protocolState.AtBlockID(receipt.ExecutionResult.BlockID).Identity(receipt.ExecutorID)
+		if err != nil {
+			return nil, fmt.Errorf("could not find executor identity")
+		}
+
+		// verify spock
+		verified, err := crypto.SPOCKVerify(approver.StakingPubKey, approval.Body.Spock, executor.StakingPubKey, receipt.Spocks[approval.Body.ChunkIndex])
+		if err != nil {
+			return nil, fmt.Errorf("could not verify spocks: %w", err)
+		}
+		if verified {
+			// map doesnt exist create it
+			if _, ok = v.approvals[approval.Body.ExecutionResultID]; !ok {
+				v.approvals[approval.Body.ExecutionResultID] = make(map[flow.Identifier]*flow.ExecutionReceipt)
+			}
+
+			// cache result for future verify calls for this approval
+			v.approvals[approval.Body.ExecutionResultID][approval.ID()] = receipt
+
+			return receipt, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// ClearReceipts clears all receipts for a specific resultID
+func (v *Verifier) ClearReceipts(resultID flow.Identifier) bool {
+	v.Lock()
+	defer v.Unlock()
+
+	// check if entry exists
+	_, ok := v.receipts[resultID]
+	if !ok {
+		return false
+	}
+	_, ok = v.approvals[resultID]
+	if !ok {
+		return false
+	}
+
+	// clear receipts and approval cache
+	delete(v.receipts, resultID)
+	delete(v.approvals, resultID)
+
+	return true
+}
+
+func (v *Verifier) matchReceipt(receipt *flow.ExecutionReceipt) (bool, error) {
+	unmatchedReceipts := v.receipts[receipt.ExecutionResult.ID()]
+
+	// get idenitity of candidate receipt
+	identity, err := v.protocolState.AtBlockID(receipt.ExecutionResult.BlockID).Identity(receipt.ExecutorID)
+	if err != nil {
+		if protocol.IsIdentityNotFound(err) {
+			return false, engine.NewInvalidInputErrorf("could not get executor identity: %w", err)
+		}
+		// unknown exception
+		return false, fmt.Errorf("could not get executor identity: %w", err)
+	}
+
+	// for through each of the receipts to check for possible match of spocks
+	// all the spocks in a receipt will have to match in order to be counted as
+	// matched
+MatchingReceiptsLoop:
+	for _, u := range unmatchedReceipts {
+		// we check if the receipt chunks length matches, if they dont match we return
+		// false by default
+		if u.ExecutionResult.Chunks.Len() != receipt.ExecutionResult.Chunks.Len() {
+			return false, nil
```

Comment:
Suggest adding an error when we return on an unhappy path (`false` in case). Otherwise would be hard to diagnose runtime issues at protocol-level.